include/envoy/server/guarddog_config.h

```diff
@@ -0,0 +1,45 @@
+#pragma once
+
+#include "envoy/api/api.h"
+#include "envoy/common/pure.h"
+#include "envoy/config/bootstrap/v3/bootstrap.pb.h"
+#include "envoy/config/typed_config.h"
+#include "envoy/protobuf/message_validator.h"
+#include "envoy/server/guarddog.h"
+
+#include "common/protobuf/protobuf.h"
+
+namespace Envoy {
+namespace Server {
+namespace Configuration {
+
+struct GuardDogActionFactoryContext {
+  // TODO(kbaichoo): Fill this with useful context that might be used
+  int foo;
+};
+
+/**
+ * Implemented by each custom GuardDogAction and registered via Registry::registerFactory()
+ * or the convenience class RegisterFactory.
+ */
+class GuardDogActionFactory : public Config::TypedFactory {
+public:
+  ~GuardDogActionFactory() override = default;
+
+  /**
+   * Creates a particular GuardDog Action factory implementation.
+   *
+   * @param config supplies the configuration for the action.
+   * @param context supplies the GuardDog Action's context.
+   * @return GuardDogActionCb the callback for the action.
+   */
+  virtual GuardDogActionCb createGuardDogActionFromProto(
```

Comment:
The watchdog actions may want to have some internal state to track things like time since the last time they executed a mega miss action, or if they are still performing some background operation triggered by a previous miss action.  I think it is probably fine to hide these things behind a callback since we can in theory create arbitrary shared_ptr objects that the callback can refer to internally, but still - would it make sense to have this action creator return an instance that implements some interface?  something like:

class GuardDogAction {
public:
  virtual ~GuardDogAction() = default;
  virtual void run(envoy::config::bootstrap::v3::Watchdog::WatchdogAction::WatchdogEvent,
    ::std::vector<::std::pair<Thread::ThreadId, MonotonicTime>>, MonotonicTime) = PURE;
};

using GuardDogActionPtr = std::unique_ptr<GuardDogAction>;
virtual GuardDogActionPtr createGuardDogActionFromProto(